Deploy cross-chain EVM bridge using Chainsafe (#675)
ava-labs/avalanche-docs: docs/build/tutorials/platform/deploying-cross-chain-evm-bridge.md

# Deploying Cross-Chain EVM <-> EVM Bridge

## Introduction

In this tutorial, we will be building a bridge between **WAGMI** and **Fuji**. This bridge will help us to transfer native **WGM** coin wrapped into **wWGM** back and forth from the WAGMI chain to the Fuji chain. Using this guide, you can deploy a bridge between any EVM-based chains for any ERC20 tokens.

The wrapped version of a native coin is its pegged ERC20 representation. Wrapping it with the ERC20 standard makes certain processes like delegated transactions much easier. You can easily get wrapped tokens by sending the native coin to the wrapped token contract address.

> WAGMI is an independent EVM-based test chain deployed on a custom subnet on the Avalanche network.

We will be using **Chainsafe**'s bridge repository, to easily set up a robust and secure bridge.

## Workflow of the Bridge

> Two chains are not interconnected. But we could communicate between the chains through off-chain relayers. Relayers watch for events (by polling blocks) on one chain and perform necessary action using those events on the other chain. This way we can also perform bridging of tokens from one chain to the other chain through the use of smart contracts.

Reviewer: ```suggestion
WAGMI and Fuji chains are not interconnected by default, however, we could make them communicate through off-chain relayers. Relayers watch for events (by polling blocks) on one chain and perform necessary action using those events on the other chain. This way we can also perform bridging of tokens from one chain to the other chain through the use of smart contracts.
```

Here is the basic high-level workflow of the bridge -

> * User deposits token on the Bridge contract

Reviewer: ```suggestion
* Users deposit token on the Bridge contract
```

* Bridge contract asks Handler contract to perform deposit action
* Handler contract **locks** the deposited token in the token safe
* Bridge contract emits `Deposit` event
* Relayer receives the `Deposit` event from the source chain
* Relayer creates a voting proposal on the destination chain to mint a new token
* After threshold relayer votes, the proposal is executed
* Tokens are **minted** to the recipient's address

Bridging tokens from source to destination chain involves the **lock and mint** approach. Whereas bridging tokens from destination to source chain involves **burn and release** approach. We cannot mint and burn tokens that we do not control. Therefore we lock them in the token safe on the source chain. And mint the corresponding token (which we will deploy and hence control) on the destination chain.

![](/img/chainsafe-bridge-1-workflow.png)

## Requirements

These are the requirement to follow this tutorial -

> * Add WAGMI and Fuji chain on the MetaMask network

Reviewer: Link to WAGMI (https://0bb4a0f8.avalanche-docs.pages.dev/build/tutorials/platform/subnets/wagmi#adding-wagmi-to-metamask) and Fuji (https://0bb4a0f8.avalanche-docs.pages.dev/build/tutorials/platform/fuji-workflow#set-up-fuji-network-on-metamask-optional) using relative path, not direct URL.

DST_ADDR="<Your address on Fuji>"
DST_PK="<your private key on Fuji>"

SRC_TOKEN="0x3Ee7094DADda15810F191DD6AcF7E4FFa37571e4"
RESOURCE_ID="0x00"
```

* `SRC_ADDR` and `DST_ADDR` are the addresses that will deploy bridge contracts and will act as a relayer.
* `SRC_TOKEN` is the token that we want to bridge. Here is the address of the wrapped ERC20 version of the WGM coin aka wWGM.
* `RESOURCE_ID` could be anything. It identifies our bridged ERC20 tokens on both sides (WAGMI and Fuji).

Every time we make changes to these config variables, we have to update our bash environment. Run the following command according to the relative location of the file. These variables are temporary and are only there in the current terminal session, and will be flushed, once the session is over. Make sure to load these environment variables anywhere you will using them in the bash commands (like `$SRC_GATEWAY` or `$SRC_ADDR`)

```bash
source ./configVars
```

## Setting Up Source Chain

We need to set up our source chain as follows -

* Deploy Bridge and Handler contract with `$SRC_ADDR` as default and only relayer
* Register the `wWGM` token as a resource on the bridge

> ### Deploy contracts

Reviewer 2: You need to note in **BOLD** that the deployer of these contracts retains admin rights that can be used to mint arbitrary amounts of the token. I'm not sure if there is a way to remove this right but if so we probably want to document it.

The command-line tool `cb-sol-cli` will help us to deploy the contracts. Run the following command in the terminal session where the config vars are loaded. It will add `SRC_ADDR` as the default relayer for relaying events from the Wagmi chain (source) to the Fuji chain (destination).

> ```bash

Reviewer 2: You should also note in **BOLD** that you should choose a very large expiry value. Expired proposals were the only time manual intervention was required on the bridge. Once a proposal expiry config is set, it cannot be modified.

```bash
echo "{
  \"chains\": [
    {
      \"name\": \"WAGMI\",
      \"type\": \"ethereum\",
      \"id\": \"0\",
      \"endpoint\": \"$SRC_GATEWAY\",
      \"from\": \"$SRC_ADDR\",
      \"opts\": {
        \"bridge\": \"$SRC_BRIDGE\",
        \"erc20Handler\": \"$SRC_HANDLER\",
        \"genericHandler\": \"$SRC_HANDLER\",
        \"gasLimit\": \"1000000\",
        \"maxGasPrice\": \"25000000000\",
        \"http\": \"true\",
        \"blockConfirmations\":\"0\"
      }
    },
    {
      \"name\": \"Fuji\",
      \"type\": \"ethereum\",
      \"id\": \"1\",
      \"endpoint\": \"$DST_GATEWAY\",
      \"from\": \"$DST_ADDR\",
      \"opts\": {
        \"bridge\": \"$DST_BRIDGE\",
        \"erc20Handler\": \"$DST_HANDLER\",
        \"genericHandler\": \"$DST_HANDLER\",
        \"gasLimit\": \"1000000\",
        \"maxGasPrice\": \"25000000000\",
        \"http\": \"true\",
        \"blockConfirmations\":\"0\"
      }
    }
  ]
}" >> config.json
```
Check and confirm the details in the `config.json` file.

> > In the above command, you can see that `blockConfirmations` is set to `0`. This will work well for networks like Avalanche because the block is confirmed once it's committed. Unlike other chains such as Ethereum, which requires 6-10 block confirmations. Therefore, use this configuration with caution, depending on the type of chain you are using.

Reviewer 2: You should say 20-30 block confirmations for ETH.

Here is the basic high-level workflow of the bridge -

* Users deposit token on the Bridge contract
* Bridge contract asks Handler contract to perform deposit action
* Handler contract **locks** the deposited token in the token safe
* Bridge contract emits `Deposit` event
* Relayer receives the `Deposit` event from the source chain
* Relayer creates a voting proposal on the destination chain to mint a new token
* After threshold relayer votes, the proposal is executed
* Tokens are **minted** to the recipient's address

Bridging tokens from source to destination chain involves the **lock and mint** approach. Whereas bridging tokens from destination to source chain involves **burn and release** approach. We cannot mint and burn tokens that we do not control. Therefore we lock them in the token safe on the source chain. And mint the corresponding token (which we will deploy and hence control) on the destination chain.

![](/img/chainsafe-bridge-1-workflow.png)

## Requirements

These are the requirement to follow this tutorial -

> * Add [WAGMI](./subnets/wagmi) and [Fuji](./fuji-workflow) chain on the MetaMask network

Reviewer: ```suggestion
* Add [WAGMI](./subnets/wagmi#adding-wagmi-to-metamask) and [Fuji](./fuji-workflow#set-up-fuji-network-on-metamask-optional) chain on the MetaMask network
```
